applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.h

```diff
@@ -0,0 +1,599 @@
+			int num_of_deriv = stress_DV_deriv.size2();
+			double stress_DV_deriv_value = 0.0;
+
+			if(rResponseGradient.size() != stress_DV_deriv.size1())
+				rResponseGradient.resize(stress_DV_deriv.size1(), false);
+			if(rResponseGradient.size() != rDerivativesMatrix.size1())
+				KRATOS_ERROR << "Size of partial stress design variable derivative does not fit!" << std::endl;
+
+			for (int dv_it = 0 ; dv_it < num_of_DV; dv_it++)
+			{
+				if(m_stress_treatment == "mean")
+				{
+					for(int GP_it = 0; GP_it < num_of_deriv; GP_it++)
+						stress_DV_deriv_value += stress_DV_deriv(dv_it, GP_it);
+
+					stress_DV_deriv_value /= num_of_deriv;
+				}
+				else if(m_stress_treatment == "GP")
+				{
+					if(num_of_deriv >= m_id_of_location)
+						stress_DV_deriv_value = stress_DV_deriv(dv_it, (m_id_of_location-1));
+					else
+						KRATOS_ERROR << "Chosen Gauss-Point is not availible. Chose 'stress_location' between 1 and " <<
+									num_of_deriv  << "!"<< std::endl;
+				}
+				else if(m_stress_treatment == "node")
+				{
+					if(num_of_deriv >= m_id_of_location)
+						stress_DV_deriv_value = stress_DV_deriv(dv_it, (m_id_of_location-1));
+					else
+						KRATOS_ERROR << "Chosen node is not availible. The element has only " <<
+									num_of_deriv  << " nodes."<< std::endl;
+				}
+				rResponseGradient[dv_it] = stress_DV_deriv_value;
+				//std::cout << ("partiall derivative (shape) = ") << stress_DV_deriv_value << std::endl;
+				stress_DV_deriv_value = 0.0;
+			}
+
+			rAdjointElem.SetValue(DESIGN_VARIABLE_NAME, "");
+		}
+		else
+		{
+			if (rResponseGradient.size() != rDerivativesMatrix.size1())
+          			rResponseGradient.resize(rDerivativesMatrix.size1(), false);
+			rResponseGradient.clear();
+		}
+
+		KRATOS_CATCH("");
+	}
+
+	// ==============================================================================
+	void CalculateSensitivityGradient(Condition& rAdjointCondition,
+                                      const Variable<array_1d<double,3>>& rVariable,
+                                      const Matrix& rDerivativesMatrix,
+                                      Vector& rResponseGradient,
+                                      ProcessInfo& rProcessInfo)
+    {
+		KRATOS_TRY;
+
+		if(rResponseGradient.size() != rDerivativesMatrix.size1())
+          	rResponseGradient.resize(rDerivativesMatrix.size1(), false);
+		rResponseGradient.clear();
+
+		KRATOS_CATCH("");
+	}
+
+	// ==============================================================================
+
+	///@}
+	///@name Protected  Access
+	///@{
+
+	///@}
+	///@name Protected Inquiry
+	///@{
+
+	///@}
+	///@name Protected LifeCycle
+	///@{
+
+	///@}
+
+private:
+	///@name Static Member Variables
+	///@{
+
+	///@}
+	///@name Member Variables
+	///@{
+
+	double m_stress_value;
+	//double m_initial_value;
+	//bool m_initial_value_defined;
+	unsigned int m_id_of_traced_element;
+	int m_id_of_location;
+	Element::Pointer m_traced_pElement;
+	std::string m_traced_stress_type;
```

Comment:
it would be really nice to use an enum for this. It would help a lot with performance